runtime/jcl/common/jclcinit.c

```diff
@@ -66,127 +66,120 @@ jint computeFullVersionString(J9JavaVM* vm)
-	if (NULL != osname) {
-		strcat(fullversion, osname);
-		strcat(fullversion, " ");
-		strcat(vminfo, osname);
-		strcat(vminfo, " ");
-	}
 	osarch = j9sysinfo_get_CPU_architecture();
-	strcat(fullversion, osarch);
-	strcat(vminfo, osarch);
 
 #ifdef J9VM_ENV_DATA64
-	strcat(fullversion, "-64 ");
-	strcat(vminfo, "-64 ");
-#ifdef J9VM_GC_COMPRESSED_POINTERS
-	strcat(fullversion, "Compressed References ");
-	strcat(vminfo, "Compressed References ");
-#endif
-#else
-#if defined(J9ZOS390) || defined(S390)
-	strcat(fullversion, "-31 ");
-	strcat(vminfo, "-31 ");
+	#ifdef J9VM_GC_COMPRESSED_POINTERS
+		#define MEM_INFO "-64 Compressed References "
+	#else
+		#define MEM_INFO "-64 "
+	#endif
 #else
-	strcat(fullversion, "-32 ");
-	strcat(vminfo, "-32 ");
+	#if defined(J9ZOS390) || defined(S390)
+		#define MEM_INFO "-31 "
+	#else
+		#define MEM_INFO "-32 "
+	#endif
 #endif
-#endif
-
-	strcat(fullversion, EsBuildVersionString);
-	strcat(vminfo, EsBuildVersionString);
-
-#if defined(J9VM_INTERP_NATIVE_SUPPORT)
-	jitConfig = vm->jitConfig;
-	if (NULL != jitConfig) {
-		if (jitConfig->runtimeFlags & J9JIT_JIT_ATTACHED) {
-			jitEnabled = 1;
-		}
-		if (jitConfig->runtimeFlags & J9JIT_AOT_ATTACHED) {
-			aotEnabled = 1;
-		}
-	}
-	strcat(fullversion, " (JIT ");
-	strcat(fullversion, jitEnabled ? "en" : "dis");
-	strcat(vminfo, " (JIT ");
-	strcat(vminfo, jitEnabled ? "en" : "dis");
-	strcat(fullversion, "abled, AOT ");
-	strcat(fullversion, aotEnabled ? "en" : "dis");
-	strcat(vminfo, "abled, AOT ");
-	strcat(vminfo, aotEnabled ? "en" : "dis");
-	strcat(fullversion, "abled)\nOpenJ9   - ");
-	strcat(vminfo, "abled)\nOpenJ9   - ");
-#endif /* J9VM_INTERP_NATIVE_SUPPORT */
-
-	vmVersion = J9VM_VERSION_STRING;
-	strcat(fullversion, vmVersion);
-	strcat(vminfo, vmVersion);
 
 #if defined(J9VM_GC_MODRON_GC)
-	gcVersion = OMR_VERSION_STRING;
-	strcat(fullversion, "\nOMR      - ");
-	strcat(fullversion, gcVersion);
-	strcat(vminfo, "\nOMR      - ");
-	strcat(vminfo, gcVersion);
+	#define OMR_INFO "\nOMR      - " OMR_VERSION_STRING
+#else
+	#define OMR_INFO ""
 #endif /* J9VM_GC_MODRON_GC */
 
 #if defined(VENDOR_SHORT_NAME) && defined(VENDOR_SHA)
-	strcat(fullversion, "\n" VENDOR_SHORT_NAME "      - " VENDOR_SHA);
-	strcat(vminfo, "\n" VENDOR_SHORT_NAME "      - " VENDOR_SHA);
+	#define VENDOR_INFO "\n" VENDOR_SHORT_NAME "      - " VENDOR_SHA
+#else
+	#define VENDOR_INFO ""
 #endif /* VENDOR_SHORT_NAME && VENDOR_SHA */
 
+	if (bufferSize <= j9str_printf(PORTLIB, fullversion, bufferSize + 1,
+			"JRE %s IBM J9 %s %s %s " MEM_INFO "%s" JIT_INFO J9VM_VERSION_STRING OMR_INFO VENDOR_INFO,
+			j2se_version_info,
+			EsVersionString,
+			(NULL != osname ? osname : " "),
+			osarch,
+			EsBuildVersionString,
+			jitEnabled,
+			aotEnabled)) {
+		j9tty_err_printf(PORTLIB, "\n%s - %d: %s: Error: Java full version string exceeds buffer size\n" __FILE__, __LINE__, __FUNCTION__);
+		return JNI_ERR;
+	}
+
+	j9str_printf(PORTLIB, vminfo, bufferSize,
```

Comment:
Please check for overflow here too. Today we might argue that this always builds a smaller string than above, but that may not always be the case and the check is nearly free.